### PR TITLE
allow method idx + 1 index

### LINKED ISSUE #6417: ZUntaggedIndex error during binary emit
Occasionally when submitting PRs we get the error described here: https://github.com/Microsoft/visualfsharp/pull/6415#issuecomment-479176337

> The FSC compiler from the .NET SDK using the bootstrap is hitting a bug on Linux and macOS as of commit #97d7a02: 
```
    13>FSC : error FS2014: A problem occurred writing the binary '/Users/vsts/agent/2.149.2/work/1/s/fcs/../artifacts/obj/fcs/net461/FSharp.Compiler.Service.dll': EmitZUntaggedIndex: too big for small address or simple index [/Users/vsts/agent/2.149.2/work/1/s/fcs/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj]
```
> This bug has cropped up now and then.  The current compiler code has some extra instrumentation for this error to help us diagnose it, but compiling using the exact commit above might help us understand it too

### PR DESCRIPTION
**Pull request by dsyme**
Fixes #6417 , which happens when we write a binary with exactly 65535 methods.

I checked the binary writer code in Mono.Cecil and our computations for sizes are correct, as is our token writing code.  The issue happens because an empty type definition (no methods) emits a "method table idx + 1" value, but that is expected.

https://github.com/jbevain/cecil/blob/ab5075b3a885ffd5bee5d347b79ce71aabeb32d1/Mono.Cecil.Metadata/Buffers.cs#L99

I'm testing the compiler referenced in the bug by hand

### CODE REVIEW

File: src/absil/ilwrite.fs [hunk 1]
```diff
@@ -87,16 +87,13 @@ type ByteBuffer with
...
-#else
-            let trace = (new System.Diagnostics.StackTrace()).ToString()
-#endif
-            failwithf "EmitZUntaggedIndex: index into table '%d' is too big for small address or simple index, idx = %d, big = %A, size of table = %d, stack = %s" nm idx big sz trace
-        else buf.EmitInt32AsUInt16 idx
+        else
+            // idx 0x10000 is allowed for method table idx + 1 for just beyond last index of method table
+            if idx > 0x10000 then 
+                System.Diagnostics.Debug.Assert (false, "EmitZUntaggedIndex: too big for small address or simple index")
+            buf.EmitInt32AsUInt16 idx
```
Comment (KevinRansom) [inline src/absil/ilwrite.fs]: So when an idx of 0x10000 is passed in, 0x0000 is written to buffer by this, which is identical to passing an idx of 0x0000.  Is that actually what we want?

**Comment by dsyme** [comment]
> So when an idx of 0x10000 is passed in, 0x0000 is written to buffer by this, which is identical to passing an idx of 0x0000. Is that actually what we want?

Yeah it is weird isn't it.  

The above is certainly what Mono.Cecil does, at least as far as I can tell.  We can check a few other binary writers too, e.g. create a binary with exactly 65535 methods using ILASM.  

I was hoping to verify by simply executing but we are currently hitting this problem on small variations of FSHarp.Compiler.Service.dll and executing that's a little awkward except in tests.  Also testing it that way isn't really good enough is it.

AFAICS the `end+1` index of 0x10000 gets used for the "end index + 1" for the methods, fields and properties of a TypeDef.  This corresponds to these lines when reading the information back in:
```
           let ((flags, nameIdx, namespaceIdx, extendsIdx, fieldsIdx, methodsIdx) as info) = seekReadTypeDefRow ctxt idx
           let nm = readBlobHeapAsTypeName ctxt (nameIdx, namespaceIdx)
           let (endFieldsIdx, endMethodsIdx) = seekReadTypeDefRowExtents ctxt info idx
```
then this:
```
           let mdefs = seekReadMethods ctxt numtypars methodsIdx endMethodsIdx
```
then this:
```
and seekReadMethods (ctxt: ILMetadataReader)  numtypars midx1 midx2 =
    mkILMethodsComputed (fun () -> 
           let mdv = ctxt.mdfile.GetView()
           [| for i = midx1 to midx2 - 1 do
                 yield seekReadMethod ctxt mdv numtypars i |])
```
Here if `midx2` is zero then the loop will simply never execute its body and we get an empty array....  It looks like this might be wrong...because we are simply failing to read any methods for a type.....   If so it's possible other readers would have similar odd behaviour....

Another approach would be to refuse to emit a binary with 65535 methods and emit a dummy.

I'll create a fake C# binary where one class has 65535 methods and compare.

**Comment by dsyme** [comment]
Actually I think the problem can only occur when 
1. there are precisely 65535 methods (or fields) and
2. the _last_ type definition has no methods (or fields)

Sure enough, the last type we're emitting is `<PrivateImplementationDetails$FSharp-Compiler-Service>`, which has no methods because there are no private implementation details for that file in the compilation.

In the ECMA 335 spec, it says that each `TypeDef` row has the following `MethodList` value:
```
 MethodList (an index into the MethodDef table; it marks the first of a continguous
run of Methods owned by this Type). The run continues to the smaller of:
1. the last row of the MethodDef table or
2. the next run of Methods, found by inspecting the MethodList of the next
row in this TypeDef table 
```
Since there is no next row in the `TypeDef` table (this is the last type definition), then the second bullet doesn't apply, and the **end** of the method range  will be the last row in the `MethodDef` table.  But what should the **start** be?  No number in the index range 0 - 65535 is valid to represent the absence of methods.  We really want `MethodDefTable.Length` (65536).  Any other index is wrong.

I actually believe this is a flaw in the .NET IL binary format.... Wow.... You simply can't represent this case.  Bang.  

Is there a bounty prize for this kind of thing after 20 years of .NET?

**Comment by jbevain** [comment]
:)

**Comment by dsyme** [comment]
@jbevain Have I got this wrong?

**Comment by jbevain** [comment]
@dsyme no I think you're spot on, I was just amused. Thinking about it this could show up elsewhere as well, I think properties and events also use ranges in a list.

**Comment by dsyme** [comment]
@jbevain Should `0` be emitted to represent the absence of methods in this case?

**Comment by dsyme** [comment]
> @dsyme no I think you're spot on, I was just amused. Thinking about it this could show up elsewhere as well, I think properties and events also use ranges in a list.

Yeah, also method parameters.  Though maybe the fact that all methods have a `return` parameter mean that's never hit.

**Comment by jbevain** [comment]
@dsyme yes I think that would work. That reminds me of an issue in Cecil that was triggered by the portable pdb support in `fsc` where we were not expecting `0` for a `*List`, but where that's actually valid to represent the absence of a list.

**Comment by dsyme** [comment]
It looks like Mono.Cecil handles a list start range of zero:

https://github.com/jbevain/cecil/blob/ab5075b3a885ffd5bee5d347b79ce71aabeb32d1/Mono.Cecil/AssemblyReader.cs#L971

So I think 0 is the right value to emit here.  But Ecma 335 just doesn't document it

**Comment by dsyme** [comment]
There is this note in ECMA 335 on a different section.  So yes, it looks like 0 is the right thing to emit.
```
18. MethodList can be null or non-null 
If MethodList is non-null, it shall index a valid row in the MethodDef table, where
valid means 1 <= row <= rowcount+1 [ERROR] 
```

**Comment by jbevain** [comment]
That was added in August 2017:

https://github.com/jbevain/cecil/commit/22b36c3dc5dcc0ddba38164904a1b3daf02151e2

So yes, not a common case, but 0 should be valid.

**Comment by dsyme** [comment]
OK, I updated the PR so our binary reader respects '0' in the start-of-list entries to mean an empty list.

I think the PR is sound now, but we should really double check the code for the .NET runtime binary reader - I'm sure it will record a zero entry as an empty list to.

**Comment by KevinRansom** [comment]
@dsyme, I think you are right that they use 0 as the start row id to indicate 0 methods, fields and properties.

GetMethodRange for a type returns the start and end row in the method table: https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs#L864

In methoddefinitionhandlecollection here: https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/HandleCollections.TypeSystem.cs#L456

Notice the count property is last - first + 1, (0 - 1 +1) = 0

**Comment by KevinRansom** [comment]
@dsyme, this looks ready to pull, are you good with it?

**Pull request merged**